### PR TITLE
Chapter 8 - Basic Login

### PR DESCRIPTION
**Pull request by EvonneNgXinYi**
Login upon signup

### CODE REVIEW

**Comment by EvonneNgXinYi** [comment]
ready

File: app/assets/stylesheets/sessions.scss [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Sessions controller here.
+// They will automatically be included in application.css.
```
Comment (ngocvt-0484) [inline app/assets/stylesheets/sessions.scss]: remove this file because you not use

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+      log_in user
+      redirect_to user
+    else
+      flash[:danger] = t "sessions.create.fail_mess_session"
+      render :new
+    end
+  end
+
+  def destroy
+  	log_out
```
Comment (ngocvt-0484) [inline app/controllers/sessions_controller.rb]:
```suggestion
   log_out
```

File: app/helpers/sessions_helper.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  # Returns the current logged-in user (if any).
+  def current_user
+    if session[:user_id]
+      @current_user ||= User.find_by id: session[:user_id] if session[:user_id]
+    end
+  end
+
+  # Returns true if the user is logged in, false otherwise.
+  def logged_in?
+    !current_user.nil?
```
Comment (ngocvt-0484) [inline app/helpers/sessions_helper.rb]: current_user.present?

File: app/models/user.rb [hunk 4]
```diff
@@ -21,4 +21,10 @@ class User < ApplicationRecord
   def downcase_email
     email.downcase!
   end
+
+  def User.digest string
+    cost = ActiveModel::SecurePassword.min_cost ? BCrypt::Engine::MIN_COST :
+      BCrypt::Engine.cost
+      BCrypt::Password.create string, cost: cost
```
Comment (ngocvt-0484) [inline app/models/user.rb:29]:
```suggestion
   BCrypt::Password.create string, cost: cost
```

File: config/routes.rb [hunk 5]
```diff
@@ -1,10 +1,15 @@
 Rails.application.routes.draw do
+  get 'sessions/new'
```
Comment (ngocvt-0484) [inline config/routes.rb]: remove this routes because you defined     post "/login", to: "sessions#new"

File: config/routes.rb [hunk 6]
```diff
@@ -1,10 +1,15 @@
...
   scope "(:locale)", locale: /en/ do
     root "static_pages#home"
     get "/help", to: "static_pages#help"
     get "/about", to: "static_pages#about"
+    get "/contact", to: "static_pages#contact"
     get "/signup", to: "users#new"
+    get "sessions/new"
```
Comment (ngocvt-0484) [inline config/routes.rb]: remove this routes because you defined post "/login", to: "sessions#new"

**Comment by EvonneNgXinYi** [comment]
ready

File: app/views/static_pages/contact.html.erb [hunk 7]
```diff
@@ -0,0 +1,6 @@
+<% provide :title, t(".contact_title") %>
+<h1><%= t ".heading" %></h1>
+<p>
+  <%= t ".contact_desc" %>
+  <%= link_to t(".ruby_on_rails_contact"), Settings.links.rails_tutorial.contact %>.
+</p>
```
Comment (saobv-2223) [inline app/views/static_pages/contact.html.erb]: image

File: app/views/sessions/new.html.erb [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+      <%= f.label :email %>
+      <%= f.email_field :email, class: "form-control" %>
+      
+      <%= f.label :password %>
+      <%= f.password_field :password, class: "form-control" %>
+
+      <%= f.submit t(".login"), class: "btn btn-primary" %>
+    <% end %>
+
+    <p><%= t ".btn_signup" %> <%= link_to t(".signup_now"), signup_path %></p>
```
Comment (saobv-2223) [inline app/views/sessions/new.html.erb]: move into form, please

Comment (EvonneNgXinYi) [reply app/views/sessions/new.html.erb]: sorry maybe i know what is “move into form" means?

**Comment by EvonneNgXinYi** [comment]
ready

File: app/views/sessions/new.html.erb (hunk 8 above)
Comment (saobv-2223) [reply app/views/sessions/new.html.erb]: yesss, that's right

File: app/helpers/sessions_helper.rb [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+  end
+
+  # Returns true if the user is logged in, false otherwise.
+  def logged_in?
+    current_user.present?
+  end
+
+  # Logs out the current user.
+  def log_out
+    session.delete(:user_id)
```
Comment (saobv-2223) [inline app/helpers/sessions_helper.rb]: remove ()

File: app/helpers/sessions_helper.rb [hunk 10]
```diff
@@ -0,0 +1,23 @@
+module SessionsHelper
+  #Log in the given user.
+  def log_in user
+    session[:user_id] = user.id
+  end
+  # Returns the current logged-in user (if any).
```
Comment (saobv-2223) [inline app/helpers/sessions_helper.rb]: remove comment

**Comment by EvonneNgXinYi** [comment]
ready

File: app/helpers/sessions_helper.rb [hunk 11]
```diff
@@ -0,0 +1,20 @@
+module SessionsHelper
+  def log_in user
+    session[:user_id] = user.id
+  end
+
+  def current_user
+    if session[:user_id]
+      @current_user ||= User.find_by id: session[:user_id] if session[:user_id]
+    end
```
Comment (saobv-2223) [inline app/helpers/sessions_helper.rb]:
```suggestion
    @current_user ||= User.find_by id: session[:user_id] if session[:user_id]
```

Comment (EvonneNgXinYi) [reply app/helpers/sessions_helper.rb]: solved

**Comment by EvonneNgXinYi** [comment]
ready

File: app/views/sessions/new.html.erb (hunk 8 above)
Comment (saobv-2223) [reply app/views/sessions/new.html.erb]: fix this comment, please

**Comment by EvonneNgXinYi** [comment]
ready

File: app/views/sessions/new.html.erb (hunk 8 above)
Comment (EvonneNgXinYi) [reply app/views/sessions/new.html.erb]: It's very sorry, my fault, my ques is : "sorry may i know what is "move in to form" means?

Comment (EvonneNgXinYi) [reply app/views/sessions/new.html.erb]: I think i find the way. Solved

File: app/models/user.rb [hunk 12]
```diff
@@ -21,4 +21,10 @@ class User < ApplicationRecord
   def downcase_email
     email.downcase!
   end
+
+  def User.digest string
```
Comment (manhnt-0960) [inline app/models/user.rb]: use syntax class << self to define class method

**Comment by EvonneNgXinYi** [comment]
ready

**Comment by ngocvt-0484** [comment]
merged

**Pull request merged**